Adopt the ServiceExtensions framework for async text input support

whsieh opened the pull request:
#### e6a7b742d24d37554294e235e5026cd9b86abc04
<pre>
Adopt the ServiceExtensions framework for async text input support
<a href="https://bugs.webkit.org/show_bug.cgi?id=266302">https://bugs.webkit.org/show_bug.cgi?id=266302</a>
<a href="https://rdar.apple.com/119568570">rdar://119568570</a>

Reviewed by Tim Horton.

Adopt `WKSE*`-prefixed objects, which are aliased to the new classes, protocols, or types and
constants exposed from the ServiceExtensions framework (if available), or otherwise aliased to the
`UIAsync*` SPI names.

No change in behavior; see below for more details.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/ios/WebEvent.mm:
(webEventType):
(isChangingKeyModifiers):
(-[WebEvent initWithKeyEvent:]):
(-[WebEvent originalKeyEvent]):
(-[WebEvent initWithUIKeyEvent:]): Deleted.
(-[WebEvent originalUIKeyEvent]): Deleted.
* Source/WebCore/platform/ios/WebEventPrivate.h:
* Source/WebCore/platform/ios/WebSEDefinitions.h: Copied from Source/WebKit/UIProcess/API/ios/WKSEDefinitions.h.
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/Shared/DocumentEditingContext.h:
* Source/WebKit/Shared/DocumentEditingContext.mm:
(WebKit::setOptionalEditingContextProperties):
(WebKit::DocumentEditingContext::toLegacyPlatformContext):
(WebKit::DocumentEditingContext::toPlatformContext):
* Source/WebKit/UIProcess/API/ios/WKSEDefinitions.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewPrivateForTestingIOS.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _updateRuntimeProtocolConformanceIfNeeded]):
(-[WKContentView textInteractionGesture:shouldBeginAtPoint:]):
(toGestureType):
(toWKSEGestureType):
(toSelectionTouch):
(toWKSESelectionTouch):
(toWKSESelectionFlags):
(toSelectionFlags):
(selectionChangedWithGesture):
(selectionChangedWithTouch):
(-[WKContentView changeSelectionWithGestureAt:withGesture:withState:]):
(-[WKContentView changeSelectionWithGestureAt:withGesture:withState:withFlags:]):
(-[WKContentView changeSelectionWithTouchAt:withSelectionTouch:baseIsStart:withFlags:]):
(-[WKContentView changeSelectionWithTouchesFrom:to:withGesture:withState:]):
(logTextInteraction):
(-[WKContentView insertTextSuggestion:]):
(shiftKeyState):
(-[WKContentView _deferKeyEventToInputMethodEditing:]):
(-[WKContentView _interpretKeyEvent:isCharEvent:]):
(-[WKContentView dataListTextSuggestions]):
(-[WKContentView setDataListTextSuggestions:]):
(-[WKContentView updateTextSuggestionsForInputDelegate]):
(-[WKContentView _provideSuggestionsToInputDelegate:]):
(-[WKContentView _provideUITextSuggestionsToInputDelegate:]):
(toWebRequest):
(-[WKContentView requestDocumentContext:completionHandler:]):

Make the completion handler here take an `NSObject`, representing either a `UIWKDocumentRequest` or
the replacement class from ServiceExtensions. This allows us to maintain both binary and source
compatibility in all the following cases:

-   Against iOS 17, which doesn&apos;t have async text input support at all, or against versions of iOS
    17 that have `UIAsyncTextInput` but not the replacement APIs in ServiceExtensions. We fall back
    to `UIWKDocumentContext` here.

-   Building/running against versions of iOS 17 that have both `UIAsyncTextInput` and the requisite
    APIs in ServiceExtensions, but have `UIKit/async_text_input` disabled by default — here, we use
    the new ServiceExtensions API objects and flags (which, for the most part, maintain binary
    compatibility), with the exception of `WKSETextDocumentRequest`. In the completion handler, if
    async text input is disabled, we return a `UIWKDocumentRequest` to maintain binary
    compatibility.

-   Lastly, building/running against versions of iOS 17 that have both `UIAsyncTextInput` and the
    requisite APIs in ServiceExtensions, and enable `UIKit/async_text_input` by default. Here, we
    return the new `WKSETextDocumentRequest`.

(-[WKContentView selectPositionAtPoint:withContextRequest:completionHandler:]):
(-[WKContentView asyncSystemInputDelegate]):
(-[WKContentView setAsyncSystemInputDelegate:]):
(-[WKContentView handleAsyncKeyEvent:withCompletionHandler:]):
(-[WKContentView replaceText:withText:options:withCompletionHandler:]):
(-[WKContentView requestTextContextForAutocorrectionWithCompletionHandler:]):

Refactor this code to use `WebKit::DocumentEditingContext` to create the document context, to avoid
having to duplicate messy logic for initializing either of the two `WKSETextDocumentRequest` types,
depending on the SDK version. Note: we don&apos;t need to consult the async text input feature flag here
and return a legacy `UIWKDocumentRequest`, since this codepath is only ever invoked in the async
text input codepath to begin with.

(-[WKContentView adjustSelection:completionHandler:]):
(-[WKContentView shiftKeyStateChangedFrom:to:]):
(-[WKContentView selectionWillChange:]):
(-[WKContentView selectionDidChange:]):
(toUIWKGestureType): Deleted.
(toUIWKSelectionTouch): Deleted.
(toUIWKSelectionFlags): Deleted.
* Source/WebKit/UIProcess/ios/WKExtendedTextInputTraits.h:
* Source/WebKit/UIProcess/ios/WKTextInteractionWrapper.h:
* Source/WebKit/UIProcess/ios/WKTextInteractionWrapper.mm:
(-[WKTextInteractionWrapper initWithView:]):
(-[WKTextInteractionWrapper selectionChangedWithGestureAt:withGesture:withState:withFlags:]):
(-[WKTextInteractionWrapper selectionChangedWithTouchAt:withSelectionTouch:withFlags:]):
(SOFT_LINK_CLASS_OPTIONAL): Deleted.
* Source/WebKit/UIProcess/ios/WebDataListSuggestionsDropdownIOS.h:
* Source/WebKit/UIProcess/ios/WebDataListSuggestionsDropdownIOS.mm:
(+[WKDataListTextSuggestion textSuggestionWithInputText:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::requestDocumentEditingContext):

Drive-by fixes: make both the completion handler and method itself take rvalue references, instead
of passing by value.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestDocumentEditingContext):

Canonical link: <a href="https://commits.webkit.org/272018@main">https://commits.webkit.org/272018@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9b1426e461f7a8a4ad2114307bdf822142a43fbc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30358 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9035 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/31970 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32865 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27469 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/31058 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/11258 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6271 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27427 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30666 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7582 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/27236 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6492 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6643 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27063 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/34204 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/27675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/27564 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32829 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6625 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4776 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30645 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8383 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7191 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/7375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7165 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/Shared/DocumentEditingContext.mm:
```diff
@@ -58,24 +82,38 @@ static inline NSRange toNSRange(DocumentEditingContext::Range range)
-        if ([platformContext respondsToSelector:@selector(setAutocorrectedRanges:)])
-            [platformContext setAutocorrectedRanges:ranges.get()];
-    }
+    return platformContext.get();
```
(Whoops — this is the cause for the mass test failures on iOS. Fixing in a moment)

aprotyas commented on Source/WebKit/Shared/DocumentEditingContext.mm:
```diff
@@ -40,7 +40,31 @@ static inline NSRange toNSRange(DocumentEditingContext::Range range)
+template <typename ContextType>
+void setOptionalEditingContextProperties(const DocumentEditingContext& context, ContextType *platformContext, OptionSet<DocumentEditingContextRequest::Options> options)
+{
+    for (auto& rect : context.textRects)
```
Can this be a const reference?

cdumez replied:
We don't normally use `const auto&` in WebKit. I'd rather we keep `auto&`.

aprotyas commented on Source/WebKit/Shared/DocumentEditingContext.mm:
```diff
@@ -58,24 +82,38 @@ static inline NSRange toNSRange(DocumentEditingContext::Range range)
     }
 
     [platformContext setSelectedRangeInMarkedText:toNSRange(selectedRangeInMarkedText)];
+    setOptionalEditingContextProperties<UIWKDocumentContext>(*this, platformContext.get(), options);
```
Is it possible to skip the template arg here because CTAD from `platformContext.get()`?

whsieh replied:
Good point! Let me give that a try..

Edit: yep, works! I'll make the change.

aprotyas reviewed (commented):
LGTM

whsieh committed:
Adopt the ServiceExtensions framework for async text input support
https://bugs.webkit.org/show_bug.cgi?id=266302
rdar://119568570

Reviewed by Tim Horton.

Adopt `WKSE*`-prefixed objects, which are aliased to the new classes, protocols, or types and
constants exposed from the ServiceExtensions framework (if available), or otherwise aliased to the
`UIAsync*` SPI names.

No change in behavior; see below for more details.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/ios/WebEvent.mm:
(webEventType):
(isChangingKeyModifiers):
(-[WebEvent initWithKeyEvent:]):
(-[WebEvent originalKeyEvent]):
(-[WebEvent initWithUIKeyEvent:]): Deleted.
(-[WebEvent originalUIKeyEvent]): Deleted.
* Source/WebCore/platform/ios/WebEventPrivate.h:
* Source/WebCore/platform/ios/WebSEDefinitions.h: Copied from Source/WebKit/UIProcess/API/ios/WKSEDefinitions.h.
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/Shared/DocumentEditingContext.h:
* Source/WebKit/Shared/DocumentEditingContext.mm:
(WebKit::setOptionalEditingContextProperties):
(WebKit::DocumentEditingContext::toLegacyPlatformContext):
(WebKit::DocumentEditingContext::toPlatformContext):
* Source/WebKit/UIProcess/API/ios/WKSEDefinitions.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewPrivateForTestingIOS.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _updateRuntimeProtocolConformanceIfNeeded]):
(-[WKContentView textInteractionGesture:shouldBeginAtPoint:]):
(toGestureType):
(toWKSEGestureType):
(toSelectionTouch):
(toWKSESelectionTouch):
(toWKSESelectionFlags):
(toSelectionFlags):
(selectionChangedWithGesture):
(selectionChangedWithTouch):
(-[WKContentView changeSelectionWithGestureAt:withGesture:withState:]):
(-[WKContentView changeSelectionWithGestureAt:withGesture:withState:withFlags:]):
(-[WKContentView changeSelectionWithTouchAt:withSelectionTouch:baseIsStart:withFlags:]):
(-[WKContentView changeSelectionWithTouchesFrom:to:withGesture:withState:]):
(logTextInteraction):
(-[WKContentView insertTextSuggestion:]):
(shiftKeyState):
(-[WKContentView _deferKeyEventToInputMethodEditing:]):
(-[WKContentView _interpretKeyEvent:isCharEvent:]):
(-[WKContentView dataListTextSuggestions]):
(-[WKContentView setDataListTextSuggestions:]):
(-[WKContentView updateTextSuggestionsForInputDelegate]):
(-[WKContentView _provideSuggestionsToInputDelegate:]):
(-[WKContentView _provideUITextSuggestionsToInputDelegate:]):
(toWebRequest):
(-[WKContentView requestDocumentContext:completionHandler:]):

Make the completion handler here take an `NSObject`, representing either a `UIWKDocumentRequest` or
the replacement class from ServiceExtensions. This allows us to maintain both binary and source
compatibility in all the following cases:

-   Against iOS 17, which doesn't have async text input support at all, or against versions of iOS
    17 that have `UIAsyncTextInput` but not the replacement APIs in ServiceExtensions. We fall back
    to `UIWKDocumentContext` here.

-   Building/running against versions of iOS 17 that have both `UIAsyncTextInput` and the requisite
    APIs in ServiceExtensions, but have `UIKit/async_text_input` disabled by default — here, we use
    the new ServiceExtensions API objects and flags (which, for the most part, maintain binary
    compatibility), with the exception of `WKSETextDocumentRequest`. In the completion handler, if
    async text input is disabled, we return a `UIWKDocumentRequest` to maintain binary
    compatibility.

-   Lastly, building/running against versions of iOS 17 that have both `UIAsyncTextInput` and the
    requisite APIs in ServiceExtensions, and enable `UIKit/async_text_input` by default. Here, we
    return the new `WKSETextDocumentRequest`.

(-[WKContentView selectPositionAtPoint:withContextRequest:completionHandler:]):
(-[WKContentView asyncSystemInputDelegate]):
(-[WKContentView setAsyncSystemInputDelegate:]):
(-[WKContentView handleAsyncKeyEvent:withCompletionHandler:]):
(-[WKContentView replaceText:withText:options:withCompletionHandler:]):
(-[WKContentView requestTextContextForAutocorrectionWithCompletionHandler:]):

Refactor this code to use `WebKit::DocumentEditingContext` to create the document context, to avoid
having to duplicate messy logic for initializing either of the two `WKSETextDocumentRequest` types,
depending on the SDK version. Note: we don't need to consult the async text input feature flag here
and return a legacy `UIWKDocumentRequest`, since this codepath is only ever invoked in the async
text input codepath to begin with.

(-[WKContentView adjustSelection:completionHandler:]):
(-[WKContentView shiftKeyStateChangedFrom:to:]):
(-[WKContentView selectionWillChange:]):
(-[WKContentView selectionDidChange:]):
(toUIWKGestureType): Deleted.
(toUIWKSelectionTouch): Deleted.
(toUIWKSelectionFlags): Deleted.
* Source/WebKit/UIProcess/ios/WKExtendedTextInputTraits.h:
* Source/WebKit/UIProcess/ios/WKTextInteractionWrapper.h:
* Source/WebKit/UIProcess/ios/WKTextInteractionWrapper.mm:
(-[WKTextInteractionWrapper initWithView:]):
(-[WKTextInteractionWrapper selectionChangedWithGestureAt:withGesture:withState:withFlags:]):
(-[WKTextInteractionWrapper selectionChangedWithTouchAt:withSelectionTouch:withFlags:]):
(SOFT_LINK_CLASS_OPTIONAL): Deleted.
* Source/WebKit/UIProcess/ios/WebDataListSuggestionsDropdownIOS.h:
* Source/WebKit/UIProcess/ios/WebDataListSuggestionsDropdownIOS.mm:
(+[WKDataListTextSuggestion textSuggestionWithInputText:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::requestDocumentEditingContext):

Drive-by fixes: make both the completion handler and method itself take rvalue references, instead
of passing by value.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::requestDocumentEditingContext):

Canonical link: https://commits.webkit.org/272018@main
